File: leetcode/0872-leaf-similar-trees/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> list1, list2;
    
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        dfs(root1, list1);
        dfs(root2, list2);
        
        if (list1.size() != list2.size())
            return false;
        
        for (int i = 0; i < list1.size(); i++) {
            if (list1[i] != list2[i])
                return false;
        }
        
        return true;
    }
    
    void dfs(TreeNode* root, vector<int>& list) {
        if (root == NULL)
            return;
        
        if (root->left == NULL && root->right == NULL)
            list.push_back(root->val);
        dfs(root->left, list);
        dfs(root->right, list);
    }
};

File: leetcode/0872-leaf-similar-trees/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: TreeNode, root2: TreeNode) -> bool:
        self.list1, self.list2 = list(), list()
        
        self.dfs(root1, self.list1)
        self.dfs(root2, self.list2)
        
        return self.list1 == self.list2
    
    def dfs(self, root: TreeNode, l: List[int]) -> None:
        if root == None:
            return
        
        if root.left == None and root.right == None:
            l.append(root.val)
        self.dfs(root.left, l)
        self.dfs(root.right, l)